database/src/main/java/com/firebase/ui/database/paging/FirebaseDataSource.java

```diff
@@ -0,0 +1,182 @@
+package com.firebase.ui.database.paging;
+
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.paging.DataSource;
+import android.arch.paging.PageKeyedDataSource;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.Query;
+import com.google.firebase.database.ValueEventListener;
+import com.google.firebase.database.annotations.NotNull;
+
+import java.util.ArrayList;
+
+/**
+ * Data source to power a {@link FirebaseRecyclerPagingAdapter}.
+ *
+ * Note: although loadInitial, loadBefore, and loadAfter are not called on the main thread by the
+ *       paging library, we treat them as if they were so that we can facilitate retry without
+ *       managing our own thread pool or requiring the user to pass us an executor.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class FirebaseDataSource<T> extends PageKeyedDataSource<String, T> {
+
+    private Query mQuery;
+    private Class<T> mClass;
+    private ArrayList<String> mKeyList;
+
+    private static final String TAG = "FirebaseDataSource";
+
+    private final MutableLiveData<LoadingState> mLoadingState = new MutableLiveData<>();
+    private final MutableLiveData<DatabaseError> mError = new MutableLiveData<>();
+    private final MutableLiveData<ArrayList<String>> mKeyLiveData = new MutableLiveData<>();
+
+    private static final String WRONG_DATA_PATH_STATUS = "WRONG_PATH";
```

Comment:
What does "wrong" mean in this case?  The error should help the developer understand what they did wrong and what they should do to fix it.

Reply:
"wrong" means if developer given reference of database path that not exists.
For e.g.
 `FirebaseDatabase().getInstance().getReference().child("list");`
If 'list' is not there in database then path is invalid or wrong.
So for such mistake, I have given "wrong".